src/D2L.CodeStyle.Analyzers/Async/Generator/SyncGenerator.cs

```diff
@@ -0,0 +1,182 @@
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace D2L.CodeStyle.Analyzers.Async.Generator;
+
+[Generator]
+internal sealed partial class SyncGenerator : IIncrementalGenerator {
+	public void Initialize( IncrementalGeneratorInitializationContext context ) {
+		// Collect all the methods we want to generate sync versions for
+		// individually (for better incremental builds)
+		IncrementalValuesProvider<(MethodDeclarationSyntax, Compilation)> methodsToGenerate =
+			context.SyntaxProvider
+			   .CreateSyntaxProvider( IsInterestingLookingMethod, ExtractSyntax )
+			   .Where( static m => m != null )
+			   .Combine( context.CompilationProvider )
+			   .WithComparerThatIgnoresCompilation()!;
+
+		// Do the generation per method
+		IncrementalValuesProvider<MethodGenerationResult> generatedMethods =
+			methodsToGenerate
+				.Select( GenerateSyncMethod )
+				// Filter out things we simply couldn't generate anything for.
+				.Where( static mr => mr.HasValue )
+				.Select( static ( mr, _ ) => mr!.Value );
+
+		// Collect and generate the output files
+		//
+		// Opportunity for better caching here -- changing one source file
+		// should only impact one output file. Currently we only generate the
+		// methods when their source file changed, but we re-collect all of
+		// that when any file with [GenerateSync] has changed.
+		IncrementalValuesProvider<FileGenerationResult> generatedFiles =
+			generatedMethods.Collect()
+				.SelectMany( GenerateFiles );
+
+		context.RegisterSourceOutput( generatedFiles, WriteFiles );
+	}
+
+	private static bool IsInterestingLookingMethod(
+		SyntaxNode node,
+		CancellationToken _
+	) {
+		if( node is not MethodDeclarationSyntax method ) {
+			return false;
+		}
+
+		// We only care about async methods
+		if( !method.Modifiers.Any( t => t.IsKind( SyntaxKind.AsyncKeyword ) ) ) {
+			return false;
+		}
+
+		// The ones we care about have to have a particular attribute -- so
+		// perk up when we see _any_ attributes (we don't have a SemanticModel
+		// in this method.)
+		return method.AttributeLists.Any();
+	}
+
+	private static MethodDeclarationSyntax? ExtractSyntax(
+		GeneratorSyntaxContext context,
+		CancellationToken cancellationToken
+	) {
+		var methodDeclaration = (MethodDeclarationSyntax)context.Node;
+
+		var attributes = methodDeclaration.AttributeLists
+			.SelectMany( al => al.Attributes );
+
+		foreach( var attribute in attributes ) {
+			// We have the semantic model now, so make sure this method has
+			// the right attribute.
+			if( IsGenerateSyncAttribute( context.SemanticModel, attribute, cancellationToken ) ) {
+				return methodDeclaration;
+			}
+		}
+
+		return null;
+	}
+
+	internal static bool IsGenerateSyncAttribute(
+		SemanticModel model,
+		AttributeSyntax attribute,
+		CancellationToken cancellationToken
+	) {
+		var attributeConstructorSymbol = model.GetSymbolInfo( attribute, cancellationToken ).Symbol as IMethodSymbol;

```

Comment:
Not sure where they use it

https://github.com/dotnet/roslyn/blob/2b7f137ebbfdf33e9eebffe87d036be392815d2b/src/Compilers/CSharp/Portable/Symbols/Source/QuickAttributeChecker.cs#L99

Reply:
I disagree, unless we measure a performance problem here.
The semantic model in this method comes "for free", it's not created on first use.

Reply:
Ran the profiler.  Looks like because of this condition:

```
!method.Modifiers.Any( t => t.IsKind( SyntaxKind.AsyncKeyword ) )
```

And there are almost no `async` methods in the LMS, we won't have an impact.  When I remove this condition and profile, it joins the list of our worst offenders:

![image](https://user-images.githubusercontent.com/12101120/152894612-6a3f3933-8344-4cb3-b5fa-b46dae6dc1b8.png)

So it won't be a problem until async gets used more.

Reply:
Using the UnqualifiedName as a comparison, the average time is 612 ticks.  

![image](https://user-images.githubusercontent.com/12101120/152895950-9a2f59b6-9b42-4c24-b8d7-6a3f52e8a298.png)

Where as before with the `GetSymbolInfo` it was an average of 2910 ticks.

Reply:
Scanned the lms.  Only 3 existing async methods with attributes;

* https://search.d2l.dev/xref/lms/lp/framework/core/D2L.LP/LP/Aws/IoT/AwsIotWebSocket.cs?r=53ed0b24#15
* https://search.d2l.dev/xref/lms/platformTools/diagnostics/D2L.LP.Diagnostics.Web/Application/AsyncTestController.cs?r=12a8200d#26
* https://search.d2l.dev/xref/lms/platformTools/diagnostics/D2L.LP.Diagnostics.Web/Application/AsyncTestController.cs?r=12a8200d#37

So all concerns gone.

Reply:
Ok, will fix in a follow-up.